packages/eslint-plugin/tests/rules/no-misused-promises.test.ts

```diff
@@ -970,5 +1005,78 @@ console.log({ ...(condition ? Promise.resolve({ key: 42 }) : {}) });
         { line: 7, messageId: 'spread' },
       ],
     },
+    {
+      code: `
+function restPromises(first: Boolean, ...callbacks: Array<() => void>): void {}
+
+restPromises(
+  true,
+  () => Promise.resolve(true),
+  () => Promise.resolve(null),
+  () => true,
+  () => Promise.resolve('Hello'),
+);
+      `,
+      errors: [
+        { line: 6, messageId: 'voidReturnArgument' },
+        { line: 7, messageId: 'voidReturnArgument' },
+        { line: 9, messageId: 'voidReturnArgument' },
+      ],
+    },
+    {
+      code: `
+type MyUnion = (() => void) | boolean;
+
+function restUnion(first: string, ...callbacks: Array<MyUnion>): void {}
+restUnion('Testing', false, () => Promise.resolve(true));
+      `,
+      errors: [{ line: 5, messageId: 'voidReturnArgument' }],
+    },
+    {
+      code: `
+function restTupleOne(first: string, ...callbacks: [() => void]): void {}
+
+function restTupleTwo(
+  first: boolean,
+  ...callbacks: [undefined, () => void, undefined]
+): void {}
+
+function restTupleFour(
+  first: number,
+  ...callbacks: [() => void, boolean, () => void, () => void]
+): void;
+
+restTupleOne('My string', () => Promise.resolve(1));
+restTupleTwo(true, undefined, () => Promise.resolve(true), undefined);
+restTupleFour(
+  1,
+  () => Promise.resolve(true),
+  false,
+  () => {},
+  () => Promise.resolve(1),
+);
```

Comment:
Nit: would you be open to splitting this test up too? Any test where there are multiple functions being declared feels to me like it should be split.

One motivation for splitting them is that we only get the first failure (of potentially many) when one fails.

I won't block on test splitting up much more 😄

Reply:
I've split this into one test per function.